linkerd/stack/src/gate.rs

```diff
@@ -145,117 +163,210 @@ where
     type Future = S::Future;
 
     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        // If the gate is shut, wait for it to open by storing a future that
-        // will complete when it opens. This ensures that waiters are notified.
-        while self.rx.is_shut() {
-            trace!(gate.open = false);
-            if !self.is_waiting {
-                let rx = self.rx.clone();
-                self.waiting.set(async move { rx.opened().await });
-                self.is_waiting = true;
+        let permit = ready!(self.poll_acquire(cx));
+        ready!(self.inner.poll_ready(cx))?;
+        self.permit = Poll::Ready(Permit(permit));
+        Poll::Ready(Ok(()))
+    }
+
+    #[inline]
+    fn call(&mut self, req: Req) -> Self::Future {
+        self.permit = match std::mem::replace(&mut self.permit, Poll::Pending) {
+            Poll::Pending => panic!("poll_ready must be called first"),
+            Poll::Ready(..) => Poll::Pending,
+        };
+
+        self.inner.call(req)
+    }
+}
+
+impl<S> Gate<S> {
+    fn poll_acquire<Req>(&mut self, cx: &mut Context<'_>) -> Poll<Option<OwnedSemaphorePermit>>
+    where
+        S: Service<Req>,
+    {
+        if !self.acquiring {
+            match self.rx.state() {
+                State::Open => return Poll::Ready(None),
+                State::Shut => {}
+                State::Limited(sem) => match sem.try_acquire_owned() {
+                    Ok(permit) => return Poll::Ready(Some(permit)),
+                    Err(TryAcquireError::NoPermits) => {}
+                    Err(TryAcquireError::Closed) => {
+                        tracing::warn!("Gate closed");
+                        return Poll::Pending;
+                    }
```

Comment:
AFAICT --- and I may be misreading, as this is a bit subtle --- it *looks* like this is a potential task leak. when the semaphore is closed (the other side of the gate is dropped), `poll_acquire` will return `Poll::Pending`  without ever having passed the `Context` to something that might wake it. If `poll_acquire` returns `Pending`, the `poll_ready` impl for `Gate` will also return `Pending` , and we've never registered our waker with something that might wake the task up again. I believe this will leak the task into the runtime as it will never terminate.

I think that if we're in the `Limited` state but the semaphore is dropped, we need to do one of two things, depending on what we want the semantics of dropping the semaphore to be:
1. return an error from `poll_acquire` that is then returned in the `poll_ready` impl for `Gate`, marking the service as permanently dead, or
2. don't return here, and continue to try re-starting the `acquire()` future, so that we get woken if the state inside the watch ever changes again.

The correct behavior depends on when the tx side will actually drop the semaphore while the state is limited.

Reply:
This could only be triggered if the other side explicitly called [`close](https://docs.rs/tokio/1.26.0/tokio/sync/struct.Semaphore.html#method.close). We probably should at least be explicit about how that case is handled.

I don't understand how this is a _task leak_, though. Whether this service is held as pending or not is entirely up to the holder. We have no idea why the holder of the `Tx` dropped the handle, but I would expect the gate to remain in whatever state it was in when the handle is lost.

Reply:
Also, just because the semaphore is closed doesn't indicate that the gate won't be updated.  So we should _probably_ go back to waiting for the rx state to change in this case.

Reply:
Yeah, I think going back to waiting for the state to change seems correct, because the semaphore closing doesn't mean the `Tx` itself has been dropped.